http-client-core/src/main/java/io/micronaut/http/client/bind/DefaultHttpClientBinderRegistry.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.client.bind;
+
+import io.micronaut.core.annotation.AnnotationMetadata;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.bind.annotation.Bindable;
+import io.micronaut.core.convert.ConversionService;
+import io.micronaut.core.naming.NameUtils;
+import io.micronaut.core.reflect.ClassUtils;
+import io.micronaut.core.reflect.ReflectionUtils;
+import io.micronaut.core.type.Argument;
+import io.micronaut.core.util.CollectionUtils;
+import io.micronaut.core.util.StringUtils;
+import io.micronaut.http.*;
+import io.micronaut.http.annotation.Body;
+import io.micronaut.http.annotation.CookieValue;
+import io.micronaut.http.annotation.Header;
+import io.micronaut.http.annotation.RequestAttribute;
+import io.micronaut.http.cookie.Cookie;
+import io.micronaut.http.cookie.Cookies;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.lang.annotation.Annotation;
+import java.util.*;
+
+/**
+ * Default implementation of {@link HttpClientBinderRegistry} that searches by
+ * annotation then by type.
+ *
+ * @author James Kleeh
+ * @since 2.1.0
+ */
+@Singleton
+@Internal
+public class DefaultHttpClientBinderRegistry implements HttpClientBinderRegistry {
+
+    private final Map<Class<? extends Annotation>, ClientArgumentRequestBinder> byAnnotation = new LinkedHashMap<>();
+    private final Map<Integer, ClientArgumentRequestBinder> byType = new LinkedHashMap<>();
+
+    /**
+     * @param conversionService The conversion service
+     * @param binders           The request argument binders
+     */
+    @Inject
+    public DefaultHttpClientBinderRegistry(ConversionService<?> conversionService,
+                                           List<ClientArgumentRequestBinder> binders) {
+        byType.put(Argument.of(HttpHeaders.class).typeHashCode(), (ClientArgumentRequestBinder<HttpHeaders>) (context, value, request) -> {
```

Comment:
Why use `Argument.of` if you are just going to call `typeHashCode()`. Isn't it the same as doing `HttpHeaders.class.hashCode()`?

Reply:
No its not

Argument.of(String.class).typeHashCode() -> -2038437531
String.class.hashCode() -> 349885916

Reply:
Seems it is just calling `Objects.hash(String.class)` which is transforming `String.class` into `Object[] { String.class }` and producing a different hash. Still not sure what the point is.